### PR TITLE
Pipeline bug fixes

### PR DESCRIPTION
**Pull request by TylerAldrich**
## Description

A ton of various bug fixes in the pipeline. Deleting a pipeline that was previously being deleted and errored caused the edges of the graph to flip, TF state was not being saved when the apply step errored, etc.

## Type of change

Please delete options that are not relevant.

- [x] Bug fix (non-breaking change which fixes an issue)

## How Has This Been Tested?

- [x] New unit tests
- [x] A ton of manual testing

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules
- [x] I have checked my code and corrected any misspellings

### CODE REVIEW

File: src/commands/destroy/datacenter.ts [hunk 1]
```diff
@@ -41,12 +40,7 @@ async function destroy_datacenter_action(options: DestroyDatacenterOptions, name
     }
   }
 
-  const confirm = options.autoApprove || (await Inputs.promptForContinuation('Are you sure you want to proceed?'));
-
-  if (!confirm) {
-    console.error('Datacenter destruction cancelled');
-    Deno.exit(1);
-  }
+  await command_helper.pipelineRenderer.confirmPipeline(pipeline, options.autoApprove);
```
Comment (TylerAldrich) [inline src/commands/destroy/datacenter.ts:43]: Note to self: This change may actually be slightly incorrect. When deleting a datacenter it wasn't showing the pipeline preview before asking you to confirm, but I think this breaks the case where there's environments to still delete so I need to double check.

File: src/pipeline/__tests__/pipeline.test.ts [hunk 2]
```diff
@@ -248,4 +328,86 @@ describe('Pipeline', () => {
     assertEquals(plannedPipeline.steps.length, 1);
     assertEquals(plannedPipeline.steps[0].action, 'update');
   });
+
+  it('pipeline edges are flipped when removing nodes and previous pipeline created/updated them', async () => {
```
Comment (davidthor) [inline src/pipeline/__tests__/pipeline.test.ts:332]: I think I understand what this test is doing, but this description is really bad. Can you make it something like:

```
it('should reverse edges when removing nodes', () => { ... })
```

File: src/pipeline/__tests__/pipeline.test.ts [hunk 3]
```diff
@@ -248,4 +328,86 @@ describe('Pipeline', () => {
...
+    assertEquals(plannedPipeline.steps[0].action, 'delete');
+    assertEquals(plannedPipeline.steps[1].action, 'delete');
+    assertEquals(plannedPipeline.steps[2].action, 'delete');
+
+    // All edges should be the reverse of the edges from the previous pipeline
+    assertEquals(plannedPipeline.edges.length, 2);
+    assertArrayIncludes(plannedPipeline.edges, [edgeAB.reverse(), edgeBC.reverse()]);
+  });
+
+  it('pipeline edges are not flipped when removing nodes and previous pipeline tried to remove nodes', async () => {
```
Comment (davidthor) [inline src/pipeline/__tests__/pipeline.test.ts:364]: Same thing as above

```
it('should NOT flip edges when removing if previous pipeline steps were already delete steps`, () => { ... })
```

**Review by davidthor — APPROVED** [review]
I'll call this ready, but I had two nits on test naming for clarity. Do with it what you see fit and then merge it in!

File: src/pipeline/__tests__/pipeline.test.ts (hunk 2 above)
Comment (TylerAldrich) [reply src/pipeline/__tests__/pipeline.test.ts]: yeah this is a much better description 😂

**Pull request merged**